src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs

```diff
@@ -406,6 +421,28 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
         {
             perRouteConventions[i](builder);
         }
+
+        if (builder is { FilterFactories.Count: > 0 } reb && action is ControllerActionDescriptor cad)
+        {
+            var routeHandlerFilters = reb.FilterFactories;
+
+            EndpointFilterDelegate del = static invocationContext =>
+            {
+                // By the time this is called, we have the cache entry
+                var controllerInvocationContext = (ControllerEndpointFilterInvocationContext)invocationContext;
+                return controllerInvocationContext.ActionDescriptor.CacheEntry!.InnerActionMethodExecutor.Execute(controllerInvocationContext);
+            };
+
+            var context = new EndpointFilterFactoryContext(cad.MethodInfo, builder.Metadata, _serviceProvider);
+
+            for (var i = routeHandlerFilters.Count - 1; i >= 0; i--)
+            {
+                var filterFactory = routeHandlerFilters[i];
+                del = filterFactory(context, del);
+            }
+
+            cad.FilterDelegate = del;
```

Comment:
How about we keep a reference to the `originalDelegate` and we did the same thing we do in RequestDelegateFactory.

https://github.com/dotnet/aspnetcore/blob/8ce81e9dac0db17415dd21eea4f02ce025f3924c/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L355-L360

```suggestion
            if (!ReferenceEquals(originalDelegate, del))
            {
                cad.FilterDelegate = del;
            }
```

Reply:
Why?

Reply:
Why not? It's easy enough, and we already do it elsewhere. I'm not demanding it.

Reply:
Yea, seems fine as long as it's easy

Reply:
https://github.com/dotnet/aspnetcore/pull/42557/commits/24a8ea3d92c682255a5006d9b03e50c1a810ddb3